{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5bf25041-c632-483b-85a7-dd7a6dfe6577",
   "metadata": {},
   "source": [
    "They have different cells with the exact same genotype matrix (polytomies), hence the sampled 1/f^2 theoretical expectation is off compared to the data?\n",
    "Note that I think some mutations could have been acquired in vitro."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96de2def-8d71-40b0-8464-cd0c88462a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from typing import List, Set\n",
    "from bimuma import donor, summary\n",
    "\n",
    "CELL_TYPES = pd.read_excel(\n",
    "    \"/mnt/c/Users/fra_t/Documents/PhD/hsc/chapman2021/41586_2021_3548_MOESM18_ESM.xlsx\",\n",
    "    sheet_name=0,\n",
    "    usecols=[0, 1],\n",
    ")\n",
    "CELL_TYPES[\"Sample\"] += \"_hum\"\n",
    "CELL_TYPES.Cell_type = CELL_TYPES.Cell_type.astype(\"category\")\n",
    "CELL_TYPES.Cell_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96ace1d1-1ca8-4e85-afe7-353343e23389",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DonorChapman:\n",
    "    def __init__(self, donor_: donor.Donor, name: str):\n",
    "        self.donor = donor_\n",
    "        self.name = name\n",
    "        # bias here to take into account polytomies (cells with the same genome)\n",
    "        # 2.4 is the mutation rate estimated in the paper\n",
    "        # self.donor.sfs[1] = 2.4 * len(self.donor.cells)\n",
    "\n",
    "\n",
    "def load_donor(\n",
    "    donor_8pcw: bool, name: str, cells2keep: List[str] | None = None\n",
    ") -> DonorChapman:\n",
    "    binary_mut = read_data(donor_8pcw)\n",
    "    if cells2keep:\n",
    "        binary_mut = summary.filter_cells_from_matrix(binary_mut, cells2keep)\n",
    "        assert len(cells2keep) == binary_mut.matrix.shape[1]\n",
    "        cells = len(cells2keep)\n",
    "    print(f\"{binary_mut.polytomies} cells with the same genotypes (polytomies)\")\n",
    "    return DonorChapman(donor.Donor(binary_mut), name)\n",
    "\n",
    "\n",
    "def load_donor_14pwc(organs: str = \"all\") -> DonorChapman:\n",
    "    if organs == \"femur_1\":\n",
    "        cells2keep = load_cell_types_donor_14pwc({\"F1\"})\n",
    "    elif organs == \"femur_2\":\n",
    "        cells2keep = load_cell_types_donor_14pwc({\"F2\"})\n",
    "    elif organs == \"femurs\":\n",
    "        cells2keep = load_cell_types_donor_14pwc({\"F1\", \"F2\"})\n",
    "    elif organs == \"liver\":\n",
    "        cells2keep = load_cell_types_donor_14pwc({\"L\"})\n",
    "    elif organs == \"all\":\n",
    "        cells2keep = load_cell_types_donor_14pwc({\"F1\", \"F2\", \"L\"})\n",
    "        assert len(cells2keep) == CELL_TYPES.shape[0]\n",
    "    else:\n",
    "        raise ValueError(f\"unknown value {organs} for `organs`\")\n",
    "    return load_donor(False, \"14pwc \" + organs, cells2keep)\n",
    "\n",
    "\n",
    "def load_donor_8pwc() -> DonorChapman:\n",
    "    return load_donor(True, \"8pcw\")\n",
    "\n",
    "\n",
    "def read_data(donor_8pcw: bool = True) -> summary.BinaryMutationMatrix:\n",
    "    data = pd.read_excel(\n",
    "        \"/mnt/c/Users/fra_t/Documents/PhD/hsc/chapman2021/41586_2021_3548_MOESM8_ESM.xlsx\",\n",
    "        sheet_name=0 if donor_8pcw else 2,\n",
    "        index_col=0,\n",
    "    )\n",
    "    return summary.BinaryMutationMatrix(data)\n",
    "\n",
    "\n",
    "def load_cell_types_donor_14pwc(organs_set: Set[str]) -> List[str]:\n",
    "    \"\"\"Donor 14 pwc has cells from femur 1, 2 and liver\"\"\"\n",
    "    cells2keep = CELL_TYPES.loc[\n",
    "        CELL_TYPES.Cell_type.isin(organs_set), \"Sample\"\n",
    "    ].to_list()\n",
    "    assert len(cells2keep), f\"Cannot find {organs_set} in cell types data set\"\n",
    "    return cells2keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46e944e6-a736-49bd-a13c-3101b88a99f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "donors = [\n",
    "    load_donor_8pwc(),\n",
    "    load_donor_14pwc(\"all\"),\n",
    "    load_donor_14pwc(\"femur_1\"),\n",
    "    load_donor_14pwc(\"femur_2\"),\n",
    "    load_donor_14pwc(\"femurs\"),\n",
    "    load_donor_14pwc(\"liver\"),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfed2db9-cfc1-44ef-b088-01980972bf08",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(donors[0].donor.cells) + len(donors[1].donor.cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a05c738d-2852-4dfb-b8ba-26764d85ad30",
   "metadata": {},
   "outputs": [],
   "source": [
    "for d in donors:\n",
    "    b = d.donor.burden\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "    ax.bar(x=b.keys(), height=b.values())\n",
    "    ax.set_title(f\"{d.name} with {len(d.donor.cells)} cells\")\n",
    "    ax.set_ylabel(\"Counts\")\n",
    "    ax.set_xlabel(\"Number of SNVs in cells\")\n",
    "    plt.show()\n",
    "\n",
    "    sfs = d.donor.sfs\n",
    "    # sfs_1_over_f_squared = donor.donor.sfs_1_over_f_squared_corrected(500, donor.donor.cells)\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "    ax.scatter(x=sfs.keys(), y=sfs.values())\n",
    "    ax.set_ylabel(\"Number of variants\")\n",
    "    ax.set_xlabel(\"Number of cells\")\n",
    "    # ax.plot(x=sfs_1_over_f_squared.keys(), y=sfs_1_over_f_squared.values())\n",
    "    ax.set_title(f\"{d.name} with {len(d.donor.cells)} cells\")\n",
    "    ax.set_yscale(\"log\")\n",
    "    ax.set_xscale(\"log\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b6da4e-98eb-4c6d-9214-76392bbab2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "for d, m in zip([donors[0], donors[-1]], (\".\", \"x\", \"v\", \"1\", \"2\", \"3\")):\n",
    "    sfs = d.donor.sfs\n",
    "    # sfs_1_over_f_squared = donor.donor.sfs_1_over_f_squared_corrected(500, donor.donor.cells)\n",
    "    # normalise\n",
    "    max_cell, max_var = max(sfs.keys()), max(sfs.values())\n",
    "    cell_freq, var_freq = (\n",
    "        [ele / max_cell for ele in sfs.keys()],\n",
    "        [ele / max_var for ele in sfs.values()],\n",
    "    )\n",
    "    ax.scatter(x=cell_freq, y=var_freq, label=d.name, alpha=0.8, marker=m)\n",
    "ax.set_ylabel(\"Variant density\")\n",
    "ax.set_xlabel(\"Variant frequency\")\n",
    "# ax.plot(x=sfs_1_over_f_squared.keys(), y=sfs_1_over_f_squared.values())\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xscale(\"log\")\n",
    "ax.legend()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
