{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c2484ae-5330-4d48-b7c5-86792db4ba8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import datetime\n",
    "import matplotlib as mpl\n",
    "from matplotlib.lines import Line2D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc388406-4cb7-455e-bc87-39546f0fc33c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_trajectories_gene(ax, df, genes, colors):\n",
    "    for gene, c in zip(genes, colors):\n",
    "        df_per_gene = df.loc[df.Gene == gene, [\"Age_binned\", \"VAF\", \"SardID\"]]\n",
    "        for id_ in df_per_gene.SardID.unique():\n",
    "            ax.plot(\n",
    "                df_per_gene.loc[df_per_gene.SardID == id_, \"Age_binned\"],\n",
    "                df_per_gene.loc[df_per_gene.SardID == id_, \"VAF\"],\n",
    "                color=c,\n",
    "                marker=\".\",\n",
    "                alpha=0.8,\n",
    "            )\n",
    "    ax.set_xlabel(\"Age (years)\")\n",
    "    ax.legend(\n",
    "        handles=[\n",
    "            Line2D([0], [0], color=c, lw=1, marker=\".\", label=g)\n",
    "            for g, c in zip(genes, colors)\n",
    "        ],\n",
    "        frameon=True,\n",
    "        fontsize=8,\n",
    "    )\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88867d1c-75c1-4527-acb4-433d9710d4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre = pd.read_csv(\n",
    "    \"/mnt/c/Users/fra_t/Documents/PhD/hsc/fabre2022/ALLvariants_exclSynonymous_Xadj.txt\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "bins = range(50, 108, 2)\n",
    "fabre[\"is_sequencing_artefact\"] = fabre.VAF < 0.02\n",
    "tot_donors = fabre.SardID.unique().shape[0]\n",
    "tot_donors_valid = fabre.loc[~fabre.is_sequencing_artefact, \"SardID\"].unique().shape[0]\n",
    "print(\n",
    "    f\"{tot_donors_valid} donors with valid variant over a total of {tot_donors} donors\"\n",
    ")\n",
    "fabre[\"Age_binned\"] = (\n",
    "    pd.cut(fabre.Age, bins=bins).map(lambda ele: ele.mid + 1).astype(int)\n",
    ")\n",
    "fabre.Age = fabre.Age.map(round).astype(int)\n",
    "fabre[\"Type\"] = fabre.Type.fillna(\"Unknown\")\n",
    "fabre.Type = fabre.Type.astype(\"category\")\n",
    "print(f\"{fabre.Gene.unique().shape[0]} genes\")\n",
    "print(f\"[{fabre.Age.min()}-{fabre.Age.max()}] age interval\")\n",
    "fabre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53ff52b9-6a7c-4006-bb6d-b137c6eca396",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre.Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3379748-946e-40d2-9825-267043c03ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre[fabre.Type == \"frameshift_deletion\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e298966d-2a73-414a-9178-60d7031af0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre[~fabre.is_sequencing_artefact].Age.hist(bins=bins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d29da9f-5e0a-4ddc-ae2a-e4348a14ca43",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre.loc[~fabre.is_sequencing_artefact, \"VAF\"].hist(bins=np.arange(0, 0.5, 0.02))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2a3d832-1d3f-498f-9cd9-85deec56c595",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre.loc[~fabre.is_sequencing_artefact, [\"Gene\", \"SardID\", \"VAF\"]].groupby(\n",
    "    [\"SardID\", \"Gene\"]\n",
    ").count().sort_values(by=\"VAF\", ascending=False).reset_index().Gene.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70699567-0e96-4d68-9a43-27db6d3dbb1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.lineplot(\n",
    "    fabre[~fabre.is_sequencing_artefact],\n",
    "    x=\"Age_binned\",\n",
    "    y=\"VAF\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4923f464-7c37-4101-a225-875ecc3ddc45",
   "metadata": {},
   "outputs": [],
   "source": [
    "largest_clones = (\n",
    "    fabre.loc[~fabre.is_sequencing_artefact, [\"Age\", \"SardID\", \"VAF\"]]\n",
    "    .groupby([\"SardID\", \"Age\"])\n",
    "    .max()\n",
    "    .dropna()\n",
    "    .reset_index()\n",
    ")\n",
    "largest_clones[\"is_largest_clone\"] = True\n",
    "shape_bf_merge = fabre.shape\n",
    "fabre = pd.merge(\n",
    "    left=fabre,\n",
    "    right=largest_clones,\n",
    "    how=\"left\",\n",
    "    on=[\"SardID\", \"Age\", \"VAF\"],\n",
    "    validate=\"many_to_one\",\n",
    ")\n",
    "print(fabre.shape)\n",
    "assert fabre.shape[0] == shape_bf_merge[0]\n",
    "fabre[\"is_largest_clone\"] = fabre.is_largest_clone.fillna(0).astype(bool)\n",
    "fabre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c7f8fb7-750c-4a02-b0ea-21fedd4795f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.lineplot(\n",
    "    fabre[(~fabre.is_sequencing_artefact) & (fabre.is_largest_clone)],\n",
    "    x=\"Age_binned\",\n",
    "    y=\"VAF\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91009797-9816-426e-87c0-4207f0d44fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for age in sorted(fabre.Age_binned.unique()):\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "    ax.hist(\n",
    "        fabre.loc[\n",
    "            (~fabre.is_sequencing_artefact)\n",
    "            & (fabre.is_largest_clone)\n",
    "            & (fabre.Age_binned == age),\n",
    "            \"VAF\",\n",
    "        ],\n",
    "        bins=np.arange(0, 0.5, 0.02),\n",
    "    )\n",
    "    ax.set_title(age)\n",
    "    ax.set_ylabel(\"counts\")\n",
    "    ax.set_xlabel(\"VAF\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b562743-1acf-4b46-babc-8e0497a20d0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# largest clone per donor, select only genes with AF >= THRSH\n",
    "THRSH = 0.1\n",
    "largest_largest_clones = fabre.loc[\n",
    "    (fabre.is_largest_clone) & (fabre.VAF > THRSH),\n",
    "    [\"VAF\", \"Age_binned\", \"SardID\", \"Gene\"],\n",
    "]\n",
    "genes = [\"ASXL1\", \"DNMT3A\", \"JAK2\", \"TET2\", \"TP53\", \"NF1\"]\n",
    "colors = mpl.colormaps[\"Dark2\"].colors[: len(genes)]\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "ax = plot_trajectories_gene(ax, largest_largest_clones, genes, colors)\n",
    "ax.set_ylabel(f\"Largest clones VAF (VAF>{THRSH})\")\n",
    "# plt.savefig(\"largest_clones_genes_trajectories.png\", dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd2b2a24-9234-4754-948b-09f897bd7869",
   "metadata": {},
   "outputs": [],
   "source": [
    "detected_clones.SardID.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d2b6632-e4c4-47fc-b983-9187f7a0401a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fabre.SardID.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a526cf5a-5a45-4e6f-a715-d9caeb1a92d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "detected_clones = (\n",
    "    fabre.loc[\n",
    "        (~fabre.is_sequencing_artefact) & (fabre.VAF >= 0.01),\n",
    "        [\"VAF\", \"Age_binned\", \"SardID\"],\n",
    "    ]\n",
    "    .groupby([\"SardID\", \"Age_binned\"])\n",
    "    .count()\n",
    ")\n",
    "detected_clones.rename(columns={\"VAF\": \"Detected clones\"}, inplace=True)\n",
    "detected_clones.reset_index(inplace=True)\n",
    "fabre = pd.merge(\n",
    "    left=fabre,\n",
    "    right=detected_clones,\n",
    "    how=\"left\",\n",
    "    on=[\"SardID\", \"Age_binned\"],\n",
    "    validate=\"many_to_one\",\n",
    ")\n",
    "fabre[\"Detected clones\"] = fabre[\"Detected clones\"].fillna(0).astype(int)\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "sns.lineplot(\n",
    "    fabre[[\"SardID\", \"Age_binned\", \"Detected clones\"]].drop_duplicates(),\n",
    "    x=\"Age_binned\",\n",
    "    y=\"Detected clones\",\n",
    "    legend=False,\n",
    "    marker=\"x\",\n",
    "    markeredgewidth=1.2,\n",
    "    markeredgecolor=\"orange\",\n",
    "    markerfacecolor=\"orange\",\n",
    "    ax=ax,\n",
    ")\n",
    "ax.set_ylabel(\"Number of expanded clones\")\n",
    "ax.set_xlabel(\"Age (years)\")\n",
    "ax.set_ylim([0, 4])\n",
    "plt.savefig(\"expanded_clones_robertson.png\", dpi=600)\n",
    "plt.show()\n",
    "detected_clones.drop(columns=\"SardID\").groupby(\"Age_binned\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e9a926f-657b-4c6b-a7a0-764c4a762bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "today = datetime.date.today().strftime(\"%d_%m_%Y\")\n",
    "fabre.to_csv(f\"fabre_cleaned_{today}.csv\", index=False)\n",
    "pd.read_csv(f\"fabre_cleaned_{today}.csv\").columns"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
