{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08e14195-f8be-453f-ab91-e42e9727e91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import datetime\n",
    "import matplotlib as mpl\n",
    "from matplotlib.lines import Line2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "630d38b4-bd87-4754-a682-154f8c50c9ba",
   "metadata": {},
   "source": [
    "## Bick et al. 2020 Nature\n",
    "The problem is that we dont have age here in the table :("
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3434da-57ce-4d00-93cf-57e911273115",
   "metadata": {},
   "outputs": [],
   "source": [
    "header = pd.read_excel(\n",
    "    \"/mnt/c/Users/fra_t/Documents/PhD/hsc/bick2020/41586_2020_2819_MOESM1_ESM.xlsx\",\n",
    "    header=7,\n",
    ")\n",
    "print(header.Title.tolist())\n",
    "header"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1a4140-16a2-487f-afb0-bde8942dad8d",
   "metadata": {},
   "source": [
    "## Uddin et al. 2022 Immunity & Ageing\n",
    "Here we have age, but only 11 genes... :("
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "696217f8-1a5d-49f0-be44-f43db32edc41",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_excel(\n",
    "    \"/mnt/c/Users/fra_t/Documents/PhD/hsc/uddin2022/12979_2022_278_MOESM2_ESM.xlsx\",\n",
    "    header=4,\n",
    ")\n",
    "data.drop(columns=data.columns[0], inplace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab37f0b7-af1c-4b65-94eb-94d59a2a3293",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.VAF.hist()\n",
    "data.VAF.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73cb3de8-03bb-44cc-8664-74399dd5b977",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"Sample ID\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a7dc66-4752-45de-9e7d-0371b644e8cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3))\n",
    "sns.lineplot(\n",
    "    data[\n",
    "        (data.Annotation == \"JAK2:NM_004972:exon14:c.G1849T:p.V617F\")\n",
    "        & (data.VAF >= 0.001)\n",
    "    ],\n",
    "    x=\"Age (Blood Draw)\",\n",
    "    y=\"VAF\",\n",
    "    hue=\"Sample ID\",\n",
    "    ax=ax,\n",
    "    estimator=None,\n",
    "    # marker=\".\",\n",
    ")\n",
    "ax.legend().set_visible(False)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_ylim([0.001, 1])\n",
    "ax.set_title(\"JAK2 V617F\")\n",
    "plt.show()\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3))\n",
    "sns.lineplot(\n",
    "    data[\n",
    "        (data.Annotation == \"DNMT3A:NM_022552:exon22:c.G2580A:p.W860X\")\n",
    "        & (data.VAF >= 0.001)\n",
    "    ],\n",
    "    x=\"Age (Blood Draw)\",\n",
    "    y=\"VAF\",\n",
    "    hue=\"Sample ID\",\n",
    "    ax=ax,\n",
    "    estimator=None,\n",
    "    # marker=\".\",\n",
    ")\n",
    "ax.legend().set_visible(False)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_title(\"DNMT3A W860X\")\n",
    "ax.set_ylim([0.001, 1])\n",
    "plt.show()\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3))\n",
    "sns.lineplot(\n",
    "    data[\n",
    "        (data.Annotation == \"ASXL1:NM_015338:exon12:c.3911delT:p.F1305Lfs*145\")\n",
    "        & (data.VAF >= 0.001)\n",
    "    ],\n",
    "    x=\"Age (Blood Draw)\",\n",
    "    y=\"VAF\",\n",
    "    hue=\"Sample ID\",\n",
    "    ax=ax,\n",
    "    estimator=None,\n",
    "    # marker=\".\",\n",
    ")\n",
    "ax.legend().set_visible(False)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_title(\"ASXL1\")\n",
    "ax.set_ylim([0.001, 1])\n",
    "plt.show()\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3))\n",
    "sns.lineplot(\n",
    "    data[\n",
    "        (data.Annotation == \"DNMT3A:NM_022552:exon23:c.G2645A:p.R882H\")\n",
    "        & (data.VAF >= 0.001)\n",
    "    ],\n",
    "    x=\"Age (Blood Draw)\",\n",
    "    y=\"VAF\",\n",
    "    hue=\"Sample ID\",\n",
    "    ax=ax,\n",
    "    estimator=None,\n",
    "    # marker=\".\",\n",
    ")\n",
    "ax.legend().set_visible(False)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_title(\"DNMT3A R882H\")\n",
    "ax.set_ylim([0.001, 1])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f54faa6-04cb-42aa-9e0f-593b0e05f0d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "sns.scatterplot(data, x=\"Age (Blood Draw)\", y=\"VAF\", hue=\"Gene\", ax=ax)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_ylim([9 * 10 ** (-5), 1])\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c4ba6a94-ce7a-4823-8411-d4841383e49a",
   "metadata": {},
   "source": [
    "## Roberston et al. 2022 Nature medicine\n",
    "Get the data from [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178936), use only `1PCT` because we are interested in VAF > 0.01.\n",
    "Indeed, the other tables `2PCT` are filtered versions of the tables `1PCT`, where only variants with VAF > 0.02 are kept.\n",
    "\n",
    "Use both synonymous and non-synonymous tables.\n",
    "\n",
    "To get the age of the donors, we need to do a bit of work. In the paper, they do not discriminate between the cohort `LBC1936` and the cohort `LBC1921` but those two are different.\n",
    "To assign the average age to the correct donors, we can use the column `Participiant_ID`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "274db62b-0144-49b0-91c3-93ff8d863ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set a threshold on valid mutations (see analysis below at the protein level: `search hitchhiking`)\n",
    "VALID_VARIANT_AF = 0.02"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e756ed2-8f1f-4cf4-b93f-c6a8ffc4130e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_trajectories_gene(ax, df, genes, colors):\n",
    "    for gene, c in zip(genes, colors):\n",
    "        df_per_gene = df.loc[\n",
    "            df.PreferredSymbol == gene, [\"Age\", \"AF\", \"participant_id\"]\n",
    "        ]\n",
    "        for id_ in df_per_gene.participant_id.unique():\n",
    "            ax.plot(\n",
    "                df_per_gene.loc[df_per_gene.participant_id == id_, \"Age\"],\n",
    "                df_per_gene.loc[df_per_gene.participant_id == id_, \"AF\"],\n",
    "                color=c,\n",
    "                marker=\".\",\n",
    "                alpha=0.8,\n",
    "            )\n",
    "    ax.set_xlabel(\"Age (years)\")\n",
    "    ax.legend(\n",
    "        handles=[\n",
    "            Line2D([0], [0], color=c, lw=1, marker=\".\", label=g)\n",
    "            for g, c in zip(genes, colors)\n",
    "        ],\n",
    "        frameon=True,\n",
    "        fontsize=8,\n",
    "    )\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8358507a-df79-4969-9dd4-02c0323f88d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A) load data\n",
    "# 1. synonymous\n",
    "data_synonymous = pd.read_csv(\n",
    "    \"/mnt/c/Users/fra_t/Documents/PhD/hsc/robertson2022/GSE178936_LBC_ARCHER.1PCT_VAF.Feb22.synonymous.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "data_synonymous[\"synonymous\"] = True\n",
    "# 2. non-synonymous\n",
    "data = pd.read_csv(\n",
    "    \"/mnt/c/Users/fra_t/Documents/PhD/hsc/robertson2022/GSE178936_LBC_ARCHER.1PCT_VAF.Feb22.non-synonymous.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "data[\"synonymous\"] = False\n",
    "# 3. concat\n",
    "data = pd.concat([data, data_synonymous], axis=0)\n",
    "data.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# B) processing\n",
    "data[\"cohort\"] = data.participant_id.str.extract(r\"CHIP_(.+)_.*\").astype(\"category\")\n",
    "# waves start at timepoint 0 and increase by 3 years at every timepoint\n",
    "mapping = {i: ele for i, ele in enumerate((range(0, 3 * 5, 3)), 1)}\n",
    "data[\"Age\"] = data.wave.map(mapping)\n",
    "data.loc[data.cohort == \"LBC21\", \"Age\"] += 79\n",
    "data.loc[data.cohort == \"LBC36\", \"Age\"] += 70\n",
    "data.dropna(subset=\"HGVSp\", inplace=True)\n",
    "# drop patients with chemio (see Robertson et al. 2022 Fig1H)\n",
    "data[data.PreferredSymbol == \"JAK2\"].sort_values(by=\"AF\", ascending=False)\n",
    "# find the two participants that received the chemio based on the VAF of JAK2\n",
    "# this will remove data that have mutations on many genes such as TP53, KMT2A, DNMT3A and NOTCH1\n",
    "data.drop(\n",
    "    index=data[\n",
    "        data.participant_id.isin(\n",
    "            data[\n",
    "                (data.PreferredSymbol == \"JAK2\") & (data.AF > 0.4)\n",
    "            ].participant_id.to_list()\n",
    "        )\n",
    "    ].index,\n",
    "    inplace=True,\n",
    ")\n",
    "# drop donors at age 82 because not enough data\n",
    "print(data.Age.value_counts().sort_index())\n",
    "data.drop(index=data[data.Age == 82].index, inplace=True)\n",
    "print(\"Dropping donors age 82 because not enough data\")\n",
    "# set artefact based on the paper and the figures shown later on\n",
    "data[\"is_sequencing_artefact\"] = data.AF < VALID_VARIANT_AF\n",
    "print(\n",
    "    f\"{data.is_sequencing_artefact.sum() / data.shape[0]:.2%} of entries with VAF lower than {VALID_VARIANT_AF}\"\n",
    ")\n",
    "data[\"is_outlier\"] = data.AF_Outlier_Pvalue >= 0.05  # not sure about this\n",
    "print(f\"{data.is_outlier.sum() / data.shape[0]:.2%} percentage of outliers?\")\n",
    "print(f\"{data.PreferredSymbol.unique().shape[0]} unique genes\")\n",
    "data[\"clone frequency\"] = data.AF * 2\n",
    "print(f\"{data.participant_id.unique().shape[0]} donors\")\n",
    "print(\n",
    "    f\"{data.participant_id.unique().shape[0] - data.loc[~data.is_sequencing_artefact, 'participant_id'].unique().shape[0]} donors without any valid variant\"\n",
    ")\n",
    "print(f\"[{data.Age.min()}-{data.Age.max()}] age interval\")\n",
    "data[~data.is_sequencing_artefact].Age.hist(bins=range(50, 105, 2))\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06efcd1b-94f5-4383-b8f6-37bcb3eca910",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3))\n",
    "sns.histplot(\n",
    "    data[~data.is_sequencing_artefact], x=\"AF\", hue=\"synonymous\", ax=ax, binwidth=0.01\n",
    ")\n",
    "plt.show()\n",
    "\n",
    "# LBC21 are older donors\n",
    "data.loc[~data.is_sequencing_artefact, [\"AF\", \"cohort\"]].groupby(\n",
    "    \"cohort\", observed=False\n",
    ").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0d9e71-3b2f-4ae9-9a60-91847e72c2b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the most present genes at patient level?\n",
    "data.loc[\n",
    "    ~data.is_sequencing_artefact, [\"participant_id\", \"PreferredSymbol\", \"AF\"]\n",
    "].groupby([\"participant_id\", \"PreferredSymbol\"]).count().sort_values(\n",
    "    by=\"AF\", ascending=False\n",
    ").reset_index().PreferredSymbol.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c81c7981-6000-4059-aef8-999b41e18d15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# increasing VAF at first, then decreasing\n",
    "fig, ax = plt.subplots(1, 1, layout=\"constrained\", figsize=(4, 3))\n",
    "sns.lineplot(\n",
    "    data.loc[(~data.is_sequencing_artefact) & (~data.synonymous), [\"AF\", \"Age\"]],\n",
    "    x=\"Age\",\n",
    "    y=\"AF\",\n",
    "    marker=\".\",\n",
    "    markersize=5,\n",
    "    markeredgewidth=2,\n",
    "    markeredgecolor=\"orange\",\n",
    "    markerfacecolor=\"orange\",\n",
    ")\n",
    "data.loc[~data.is_sequencing_artefact, [\"AF\", \"Age\"]].groupby(\"Age\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76fe0494-d8e9-4732-ba61-2affe45f253b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"TET2\", \"DNMT3A\", \"JAK2\", \"NF1\"]\n",
    "colors = mpl.colormaps[\"Dark2\"].colors[: len(genes)]\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "ax = plot_trajectories_gene(\n",
    "    ax, data[(~data.is_sequencing_artefact) & (~data.synonymous)], genes, colors\n",
    ")\n",
    "ax.set_ylabel(\"Clones VAF\")\n",
    "plt.savefig(\"genes_trajectories.png\", dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "655c73b8-8321-4fbc-87d0-ef0306684884",
   "metadata": {},
   "source": [
    "### Biggest clones assuming donors are the same process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca87bc6-064c-49fd-8136-c2d7f704e20b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each donor, get the largest frequency. Then, aggregate per age to compute\n",
    "# some stats on those data\n",
    "# largest clone per patient\n",
    "largest_clones = (\n",
    "    data.loc[\n",
    "        (~data.is_sequencing_artefact) & (~data.synonymous),\n",
    "        [\"participant_id\", \"AF\", \"Age\"],\n",
    "    ]\n",
    "    .groupby([\"participant_id\", \"Age\"])\n",
    "    .max()\n",
    ").reset_index()\n",
    "# largest_clones[\"Largest clone frequency\"] = 2 * largest_clones.AF\n",
    "largest_clones[\"is_largest_clone\"] = True\n",
    "# largest_clones.drop(columns=\"AF\", inplace=True)\n",
    "shape_bf_merge = data.shape\n",
    "print(shape_bf_merge)\n",
    "data = pd.merge(\n",
    "    left=data,\n",
    "    right=largest_clones,\n",
    "    how=\"left\",\n",
    "    on=[\"participant_id\", \"Age\", \"AF\"],\n",
    "    validate=\"many_to_one\",\n",
    ")\n",
    "print(data.shape)\n",
    "assert data.shape[0] == shape_bf_merge[0]\n",
    "data[\"is_largest_clone\"] = data.is_largest_clone.fillna(0).astype(bool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ea670a-1958-4884-8e5e-f6e772b221e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# VAF in fn of age\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "sns.lineplot(\n",
    "    data.loc[data.is_largest_clone, [\"participant_id\", \"Age\", \"clone frequency\"]],\n",
    "    x=\"Age\",\n",
    "    y=\"clone frequency\",\n",
    "    errorbar=(\"ci\", 95),\n",
    "    legend=False,\n",
    "    ax=ax,\n",
    "    marker=\".\",\n",
    "    markersize=5,\n",
    "    markeredgewidth=2,\n",
    "    markeredgecolor=\"orange\",\n",
    "    markerfacecolor=\"orange\",\n",
    "    n_boot=10000,\n",
    "    seed=10,\n",
    ")\n",
    "ax.set_xlabel(\"Age (years)\")\n",
    "ax.set_ylabel(\"Largest clone frequency\")\n",
    "# ax.set_ylim([0, 0.4])\n",
    "ax.set_xlim([68, 92])\n",
    "plt.savefig(\"largest_clone_avg_robertson.png\", dpi=600)\n",
    "plt.show()\n",
    "\n",
    "largest_clones.drop(columns=[\"participant_id\", \"is_largest_clone\"]).groupby(\n",
    "    \"Age\"\n",
    ").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a45a4f8-3658-4a09-9a6e-f31de8c7deaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# are the genes from 79 different from genes at age 88?\n",
    "gene_occurrences = (\n",
    "    data.loc[data.is_largest_clone, [\"Age\", \"participant_id\", \"PreferredSymbol\"]]\n",
    "    .groupby([\"Age\", \"PreferredSymbol\"])\n",
    "    .count()\n",
    "    .reset_index()\n",
    ")\n",
    "for age in [79, 88]:\n",
    "    print(age)\n",
    "    print(\n",
    "        gene_occurrences.loc[\n",
    "            gene_occurrences.Age == age, [\"PreferredSymbol\", \"participant_id\"]\n",
    "        ].sort_values(by=\"PreferredSymbol\")\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e808c3a2-d6cf-4718-a627-c863e81e298e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for age in sorted(data.Age.unique()):\n",
    "    fig, ax = plt.subplots(1, 1, layout=\"constrained\", figsize=(4, 3))\n",
    "    data_age = data.loc[\n",
    "        (data.Age == age) & (data.is_largest_clone), [\"clone frequency\"]\n",
    "    ].drop_duplicates()\n",
    "    sns.histplot(data_age, x=\"clone frequency\", bins=np.arange(0, 1.1, 0.05), ax=ax)\n",
    "    ax.set_xlim([0, 1])\n",
    "    ax.set_title(f\"{age} y.o.\")\n",
    "    ax.set_xlabel(\"Largest clone frequency\")\n",
    "    plt.savefig(f\"largest_clone_distr_per_timepoint_robertson_{age}years.svg\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d81686a7-f9b6-4a48-a381-0a341e0b4caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# largest clone per donor\n",
    "fig, ax = plt.subplots(1, 1, layout=\"constrained\")\n",
    "g = sns.lineplot(\n",
    "    data.loc[\n",
    "        (data.is_largest_clone) & (data.AF >= 0.06),\n",
    "        [\"AF\", \"Age\", \"participant_id\", \"PreferredSymbol\"],\n",
    "    ],\n",
    "    x=\"Age\",\n",
    "    y=\"AF\",\n",
    "    hue=\"PreferredSymbol\",\n",
    "    errorbar=(\"ci\", 95),\n",
    "    # legend=False,\n",
    "    palette=\"Dark2\",\n",
    "    ax=ax,\n",
    "    marker=\"x\",\n",
    "    markersize=3,\n",
    "    markeredgewidth=1,\n",
    "    markeredgecolor=\"blue\",\n",
    ")\n",
    "# sns.move_legend(g, \"upper left\", bbox_to_anchor=(1, 1), title=\"Gene\", fontsize=8)\n",
    "ax.set_ylim([0.05, 0.5])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be7cde3-2fe5-4b5e-abc1-36ca27dc8d19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# largest clone per donor, select only genes with AF >= 0.06\n",
    "largest_largest_clones = data.loc[\n",
    "    (data.is_largest_clone) & (data.AF > 0.0601),\n",
    "    [\"AF\", \"Age\", \"participant_id\", \"PreferredSymbol\"],\n",
    "]\n",
    "genes = largest_largest_clones.PreferredSymbol.unique()\n",
    "colors = mpl.colormaps[\"Dark2\"].colors[: len(genes)]\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "ax = plot_trajectories_gene(ax, largest_largest_clones, genes, colors)\n",
    "ax.set_ylabel(\"Largest clones VAF (VAF>0.06)\")\n",
    "plt.savefig(\"largest_clones_genes_trajectories.png\", dpi=600)\n",
    "plt.show()\n",
    "\n",
    "# largest clone per donor, select all genes\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "sns.lineplot(\n",
    "    data.loc[data.is_largest_clone, [\"AF\", \"Age\", \"participant_id\"]],\n",
    "    x=\"Age\",\n",
    "    y=\"AF\",\n",
    "    hue=\"participant_id\",\n",
    "    legend=False,\n",
    "    # palette=\"Blues\",\n",
    "    ax=ax,\n",
    "    marker=\"x\",\n",
    "    markersize=3,\n",
    "    markeredgewidth=1,\n",
    "    markeredgecolor=\"blue\",\n",
    ")\n",
    "ax.set_ylim([0, 0.5])\n",
    "ax.set_ylabel(\"Largest clones VAF\")\n",
    "ax.set_xlabel(\"Age (years)\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70a61e27-1051-44e4-935c-2d6087e34647",
   "metadata": {},
   "outputs": [],
   "source": [
    "detected_clones = (\n",
    "    data.loc[data.AF >= 0.01, [\"AF\", \"Age\", \"participant_id\"]]\n",
    "    .groupby([\"participant_id\", \"Age\"])\n",
    "    .count()\n",
    ")\n",
    "detected_clones.rename(columns={\"AF\": \"Detected clones with artefacts\"}, inplace=True)\n",
    "detected_clones.reset_index(inplace=True)\n",
    "data = pd.merge(\n",
    "    left=data,\n",
    "    right=detected_clones,\n",
    "    how=\"left\",\n",
    "    on=[\"participant_id\", \"Age\"],\n",
    "    validate=\"many_to_one\",\n",
    ")\n",
    "data[\"Detected clones with artefacts\"] = (\n",
    "    data[\"Detected clones with artefacts\"].fillna(0).astype(int)\n",
    ")\n",
    "detected_clones.groupby(\"Age\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46ee4262-34f8-4148-b8d4-fc17e17ece27",
   "metadata": {},
   "outputs": [],
   "source": [
    "detected_clones = (\n",
    "    data.loc[\n",
    "        (~data.is_sequencing_artefact) & (data.AF >= 0.01),\n",
    "        [\"AF\", \"Age\", \"participant_id\"],\n",
    "    ]\n",
    "    .groupby([\"participant_id\", \"Age\"])\n",
    "    .count()\n",
    ")\n",
    "detected_clones.rename(columns={\"AF\": \"Detected clones\"}, inplace=True)\n",
    "detected_clones.reset_index(inplace=True)\n",
    "data = pd.merge(\n",
    "    left=data,\n",
    "    right=detected_clones,\n",
    "    how=\"left\",\n",
    "    on=[\"participant_id\", \"Age\"],\n",
    "    validate=\"many_to_one\",\n",
    ")\n",
    "data[\"Detected clones\"] = data[\"Detected clones\"].fillna(0).astype(int)\n",
    "detected_clones.groupby(\"Age\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14fa73d9-7557-49cc-a267-081d8fed3d97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: I think we should drop age > 81 if we want to compare to Fig3F\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), layout=\"constrained\")\n",
    "sns.lineplot(\n",
    "    data[[\"participant_id\", \"Age\", \"Detected clones\"]].drop_duplicates(),\n",
    "    x=\"Age\",\n",
    "    y=\"Detected clones\",\n",
    "    legend=False,\n",
    "    marker=\"x\",\n",
    "    markeredgewidth=1.2,\n",
    "    markeredgecolor=\"black\",\n",
    "    ax=ax,\n",
    ")\n",
    "ax.set_ylabel(\"Number of expanded clones\")\n",
    "ax.set_xlabel(\"Age (years)\")\n",
    "ax.set_ylim([0, 3])\n",
    "plt.savefig(\"expanded_clones_robertson.png\", dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6567c9a8-6ede-408a-811f-f43836a9833b",
   "metadata": {},
   "outputs": [],
   "source": [
    "today = datetime.date.today().strftime(\"%d_%m_%Y\")\n",
    "data.to_csv(f\"robertson_cleaned_{today}.csv\", index=False)\n",
    "pd.read_csv(f\"robertson_cleaned_{today}.csv\").columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "441d2ad7-ca8e-4d47-86d5-a8ac4bd6e2c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hitchhiking: how can we distinguish between hichhikers and clones?\n",
    "# this is probably not hitchhiking otherwise we would observe that also\n",
    "# with higher AF, which is not the case.\n",
    "# Based on this we discard all variants with AF < 0.02 for the largest clones\n",
    "# genes = {\"NF1\", \"NOTCH1\", \"TP53\"}\n",
    "genes = sorted([\"KMT2A\", \"DNMT3A\", \"NOTCH1\", \"TP53\", \"NF1\", \"RAD21\"])\n",
    "for participant in data.loc[\n",
    "    data[\"Detected clones with artefacts\"] > 30, \"participant_id\"\n",
    "].unique():\n",
    "    xlims = (\n",
    "        data.loc[data[\"participant_id\"] == participant, \"Age\"].min() - 1,\n",
    "        data.loc[data[\"participant_id\"] == participant, \"Age\"].max() + 1,\n",
    "    )\n",
    "    fig, axes = plt.subplots(\n",
    "        2, 3, figsize=(6, 3.5), layout=\"constrained\", sharey=True, sharex=True\n",
    "    )\n",
    "    tmp = data.loc[\n",
    "        (data[\"participant_id\"] == participant) & (data[\"PreferredSymbol\"].isin(genes)),\n",
    "        [\"AF\", \"HGVSp\", \"Age\", \"PreferredSymbol\"],\n",
    "    ]\n",
    "    ylims = 0, tmp.AF.max() + 0.001\n",
    "    for gene, ax in zip(genes, axes.ravel()):\n",
    "        sns.lineplot(\n",
    "            data=tmp[tmp.PreferredSymbol == gene],\n",
    "            x=\"Age\",\n",
    "            y=\"AF\",\n",
    "            hue=\"HGVSp\",\n",
    "            marker=\"x\",\n",
    "            markersize=3,\n",
    "            markeredgewidth=1,\n",
    "            markeredgecolor=\"blue\",\n",
    "            legend=False,\n",
    "            ax=ax,\n",
    "        )\n",
    "        ax.set_title(gene)\n",
    "        ax.set_ylim(ylims)\n",
    "        ax.set_xlim(xlims)\n",
    "        ax.fill_between(\n",
    "            x=np.arange(xlims[0], xlims[1] + 1),\n",
    "            y1=0,\n",
    "            y2=0.01,\n",
    "            color=\"red\",\n",
    "            alpha=0.1,\n",
    "            edgecolor=\"face\",\n",
    "        )\n",
    "    fig.suptitle(participant)\n",
    "    plt.savefig(f'{str(\"_\").join(genes)}_{participant}.png', dpi=600)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58511106-a1f3-4998-9ca0-fe71f9504dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.relplot(\n",
    "    data.loc[\n",
    "        (data[\"PreferredSymbol\"] == \"JAK2\")\n",
    "        & (~data.Variant_Classification.isin([\"RNA\", \"5'UTR\"]))\n",
    "    ],\n",
    "    x=\"Age\",\n",
    "    y=\"AF\",\n",
    "    hue=\"participant_id\",\n",
    "    col=\"Variant_Classification\",\n",
    "    kind=\"line\",\n",
    "    marker=\".\",\n",
    ")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
